sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java

```diff
@@ -168,25 +219,99 @@ public MetricData collect(
       InstrumentationScopeInfo instrumentationScopeInfo,
       long startEpochNanos,
       long epochNanos) {
-    Map<Attributes, T> result;
+    if (memoryMode == REUSABLE_DATA) {
+      // Collect can not run concurrently for same reader, hence we safely assume
+      // the previous collect result has been used and done with
+      reusableResultList.forEach(v -> reusablePointsPool.returnObject(v));
+      reusableResultList.clear();
+    }
+
+    Collection<T> result;
     if (aggregationTemporality == AggregationTemporality.DELTA) {
       Map<Attributes, T> points = this.points;
       Map<Attributes, T> lastPoints = this.lastPoints;
-      lastPoints.entrySet().removeIf(entry -> !points.containsKey(entry.getKey()));
+
+      Collection<T> deltaPoints;
+      switch (memoryMode) {
+        case REUSABLE_DATA:
+          deltaPoints = reusableResultList;
+          break;
+        case IMMUTABLE_DATA:
+          deltaPoints = new ArrayList<>();
+          break;
+        default:
+          throw new IllegalStateException("Unsupported memory mode: " + memoryMode);
+      }
+
       points.forEach(
-          (k, v) -> lastPoints.compute(k, (k2, v2) -> v2 == null ? v : aggregator.diff(v2, v)));
-      result = lastPoints;
+          (k, v) -> {
+            T lastPoint = lastPoints.get(k);
+
+            T deltaPoint;
+            if (lastPoint == null) {
+              switch (memoryMode) {
+                case REUSABLE_DATA:
+                  deltaPoint = reusablePointsPool.borrowObject();
```

Comment:
Instead of borrowing an object and copying `v` to it, can we just set `deltaPoint` to be equal to `v`?

Reply:
I need those objects later since "points" become "lastPoints" at `this.lastPoints = points;` close to the end of the if block of DELTA.